draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,937 @@
+health checks with.
+
+In addition, if the task or the service unexpectedly fails, the user
+can configure the Circuit Breaker to trip after a configured number of times.
+The default value is 3 times.
+
+Whenever a Circuit Breaker trips, the worker will emit a warning log message.
+
+After a configured period of time the circuit is opened again and tasks may
+execute. The default period of time is 30 seconds with no linear or exponential
+growth.
+
+The user will configure the following properties of the Circuit Breaker:
+
+* How many times the health checks may fail before
+  the circuit breaker trips.
+* How many unexpected failures the task or service tolerates before tripping
+  the Circuit Breaker.
+* The period of time after which the circuit is yet
+  again closed. That time period may grow linearly or exponentially.
+* How many circuit breaker trips during a period of time should cause the worker
+  to produce an error log message instead of a warning log message.
+* The period of time after which the circuit breaker downgrades
+  it's log level back to warning.
+
+.. admonition:: Example
+
+  We allow 2 **Unhealthy** health checks
+  and/or 10 **Degraded** health checks in a period of 10 seconds.
+
+  If we cross that threshold, the circuit breaker trips.
+
+  The circuit will be closed again after 30 seconds. Afterwards, the task can
+  be executed again.
+
+  If 3 consequent circuit breaker trips occurred during a period of 5 minutes,
+  all circuit breaker trips will emit an error log message instead of a warning.
+
+  The circuit breaker will downgrade it's log level after 30 minutes.
+
+
+Network Resilience
+++++++++++++++++++
+
+Network Connections may fail at any time.
+In order to be network resilient we must use retries and circuit breakers on
+all outgoing and incoming connections.
+
+Worker
+------
+
+Services
+++++++++
+
+Services are long running tasks which are used by Celery to perform its internal
+operations.
+
+Some services publish messages to brokers, others consume messages from them.
+Other services are used to calculate optimal scheduling of tasks, routing,
+logging and even executing tasks.
+
+Users may create their own services as well.
+
+Tasks
++++++
+
+Tasks are short running, have a defined purpose and are triggered in response
+to messages.
+
+Celery declares some tasks for internal usage.
+
+Users will create their own tasks for their own use.
+
+Protocol
+++++++++
+
+Introduction to AMQP 1.0 Terminology
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Worker Health Checks
+++++++++++++++++++++
+
+Worker Circuit Breakers
++++++++++++++++++++++++
+
+Inbox Queue
++++++++++++
+
+Each worker declares an inbox queue in the :term:`Message Broker`.
+
+Publishers may publish messages to that queue in order to execute tasks on a
+specific worker.
+
+Celery uses the Inbox Queue to defer execution of the worker's internal tasks.
+
+While disabling the inbox queue is possible, some functionality will be lost.
+
+Publisher
```

Comment:
This section of publishers, backlogs, etc. seems like it could use a diagram of how all these pieces fit together!

Reply:
Yes. I agree.
I think we should complete and agree that each component I wrote about is a must before drawing a diagram.